Silver Mueller in RZ

LDAmorim opened the pull request:
This PR is an expansion of PR #1766 to add the option of using the **Silver Mueller** absorbing boundary condition (for `r=r_max`) when running WarpX in the RZ geometry (with the Yee solver).

The main changes were:
-  _Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp_  - in the RZ section. Where the multiplying coefficients, radial parameters and `Bt` and `Bz` were updated with values from `Bt`, `Bz`, `Er`, `Et` and `Ez` for azimuthal modes 0 and higher, with attention to real and imaginary components.

The equations used are derived in the file:
[SilverMueller.pdf](https://github.com/ECP-WarpX/WarpX/files/6221290/SilverMueller.pdf)


Tests:

- [x] RZ Gaussian beam propagation in vacuum with `self-fields` (without boost, mesh refinement)

A following PR will be dedicated to creating and adding to the repo a low resolution RZ **Silver Mueller** test with a laser pulse that propagates in the `r` direction (similar to _Examples/Tests/SilverMueller/inputs_2d_x_).
Then the corresponding `Python` analysis and `PICMI` scripts can be created and added to the repo.

LDAmorim commented on Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:
```diff
@@ -33,6 +32,88 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
+
+        // Loop over the cells
+        amrex::ParallelFor(tbr, tbt, tbz,
+            [=] AMREX_GPU_DEVICE (int /*i*/, int /*j*/, int /*k*/){},
```
I am not sure if this is the correct / best way to skip changes to `Br`.

RemiLehe replied:
That would work, but I would recommend the change below, which makes use of the fact that you can also pass only 2 boxes (and only 2 lambda functions) to `ParallelFor`:
https://amrex-codes.github.io/amrex/doxygen/namespaceamrex.html#a10d0f241ad4d9d28d20fa84630e7c476

LDAmorim replied:
Thank you Rémi! I will use the 2 boxes that is a much cleaner option.

LDAmorim commented:
The comparison of the fields that appear when a relativistic Gaussian particle beam propagates in RZ (without boost nor mesh refinement) without (left) and with this PR (right) show the attenuation of the reflected fields:

![imagem](https://user-images.githubusercontent.com/15057096/112169456-1e2ff680-8bea-11eb-85b1-df53982b50ad.png)

LDAmorim commented:
Using higher resolution,  [inputs_rz.txt](https://github.com/ECP-WarpX/WarpX/files/6191002/inputs_rz.txt), still shows the reduction of the reflected fields:

![imagem](https://user-images.githubusercontent.com/15057096/112176949-6f42e900-8bf0-11eb-816d-e726f7cb2a17.png)

RemiLehe commented on Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:
```diff
@@ -33,6 +32,88 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
+        tbz.grow(1);
+
+        // Loop over the cells
+        amrex::ParallelFor(tbr, tbt, tbz,
+            [=] AMREX_GPU_DEVICE (int /*i*/, int /*j*/, int /*k*/){},
```
```suggestion
        amrex::ParallelFor(tbt, tbz,
```

LDAmorim replied:
Thank you for the 2 box `ParallelFor` suggestion, it's much cleaner and more efficient.
I made this correction and removed `tbr` since we don't use it anymore.

RemiLehe commented on Source/FieldSolver/FiniteDifferenceSolver/ApplySilverMuellerBoundary.cpp:
```diff
@@ -33,6 +32,88 @@ void FiniteDifferenceSolver::ApplySilverMuellerBoundary (
+            },
+            [=] AMREX_GPU_DEVICE (int i, int j, int /*k*/){
+
+                Real const r = rmin + i*dr; // r on nodal point (Br is nodal in r)
```
Should this be modified because we are looking at a point for Bz?
```suggestion
                Real const r = rmin + (i+0.5_rt)*dr; // r on cell-centered point (Bz is cell-centered in r)
```

LDAmorim replied:
Thank you for catching this Rémi! Of course, you are totally correct. I just fixed it.

RemiLehe reviewed (changes requested):
Thanks a lot for this PR, and for the detailed explanation in the PDF document! This looks good! I have a few minor comments.

LDAmorim commented:
I have updated the calculations to use the correct azimuthal modes convention and corrected the code.
The low resolution `Gaussian` particle beam test seems to give OK results (reduction of space charge fields reflected back - to the left) when comparing RZ case with and without the *Silver-Mueller* correction:

![imagem](https://user-images.githubusercontent.com/15057096/112829945-c4b44580-9089-11eb-8b7d-ad829c94eda4.png)
